plugins/redis/redis.cpp

```diff
@@ -227,35 +323,34 @@ static class MainThreadCachedConnection
 {
 public :
     MainThreadCachedConnection() { }
-    ~MainThreadCachedConnection() { releaseContext(); }
+    ~MainThreadCachedConnection() { releaseAllCachedContexts(); }
 } mainThread;
-#endif
 
-Connection::Connection(ICodeContext * ctx, const char * _options, int _database, const char * password, unsigned _timeout)
-  : timeout(_timeout), serverIpPortPasswordHash(0)
+Connection::Connection(ICodeContext * ctx, const char * _options, const char * _ip, int _port, bool parseOptions, int _database, const char * password, unsigned _timeout)
+  : timeout(_timeout)
 {
-#if HIREDIS_VERSION_OK_FOR_CACHING
     serverIpPortPasswordHash = hashServerIpPortPassword(ctx, _options, password);
-#endif
-    options.set(_options, strlen(_options));
-    parseOptions(ctx, _options);
-    connect(ctx, _database, password);
-}
-Connection::Connection(ICodeContext * ctx, const char * _options, const char * _ip, int _port, unsigned _serverIpPortPasswordHash, int _database, const char * password, unsigned _timeout)
-  : timeout(_timeout), serverIpPortPasswordHash(_serverIpPortPasswordHash)
-{
-    port = _port;
     options.set(_options, strlen(_options));
-    ip.set(_ip, strlen(_ip));
+
+    if (parseOptions || !_ip || !_port)//parseOptions(true) is intended to be used when passing a valid ip & port but check just in case.
+        doParseOptions(ctx, _options);
+    else
+    {
+        port = _port;
+        ip.set(_ip, strlen(_ip));
+    }
     connect(ctx, _database, password);
 }
 void Connection::redisConnect()
 {
+    freeContext();
     if (timeout.getTimeout() == 0)
         context = ::redisConnect(ip.str(), port);
     else
     {
         int _timeLeft = (int) timeLeft();
```

Comment:
style: we tend to only use variables preceded by _ as parameters to the constructor.  Also, not sure why you are casting to int.